src/unlinkcommand.js

```diff
@@ -0,0 +1,90 @@
+/**
+ * @license Copyright (c) 2003-2016, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+import Command from '../core/command/command.js';
+import Range from '../engine/model/range.js';
+
+/**
+ * The unlink command. It is used by the {@link Link.Link link feature}.
+ *
+ * @memberOf link
+ * @extends core.command.Command
+ */
+export default class LinkCommand extends Command {
+	/**
+	 * Executes the command.
+	 *
+	 * When selection is collapsed then whole link will be unlinked. {@link engine.model.TreeWalker TreeWalker} is
+	 * looking for link bounds going forward and backward from the caret position.
+	 *
+	 * If selection is non-collapsed then only selected range will be unlinked.
+	 *
+	 * @protected
+	 */
+	_doExecute() {
+		const document = this.editor.document;
+		const selection = document.selection;
+		const ranges = selection.getRanges();
+		let rangesToUpdate = [];
+
+		document.enqueueChanges( () => {
+			// When selection is collapsed we have to find link bounds.
+			if ( selection.isCollapsed ) {
+				// Loop through each selection.
+				for ( let range of ranges ) {
+					// Get searching area.
+					const parentNodeRange = Range.createIn( range.start.parent );
+
+					// Create walker instances for going forward and backward.
+					const walker = parentNodeRange.getWalker( { startPosition: range.start } );
```

Comment:
It's fancy to use `TreeWalker` but I am afraid it's overcomplication, both from LOCs aspect and code readability.

Instead, you could just take the text node in which collapsed selection is placed and check it's siblings as long as they have same link value.

BTW. I haven't tried it, but isn't this implementation removing links in two separate blocks (i.e. paragraphs) if they are same? Consider this:
`<p>Lorem <a href="foo.html">ipsum</a><p><a href="foo.html">dolor</a> sit amet.</p>`
If so, I am not sure that it should work this way. This is another reason to just check siblings.